### DISCUSSION TITLE
v0.10 Dev Channel

### DISCUSSION THREAD

**Discussion by Blinue** [Blinue/Magpie#362] [Dev Channel]
v0.10 专注于提高易用性。

1. 重新设计 UI
2. 提供配置缩放模式的 UI
3. ~~支持在运行时修改效果参数以及保存~~
4. 支持保存特定游戏的配置
5. ~~支持 ARM64~~

为了尽快释出 v0.10，一些功能被推迟

**Comment by Blinue** [comment]
#223 暂时不会实现。ScaleModels.json 可能也会放弃，或者只用于分享缩放配置。

**Reply by bobby3510** [reply]
1

**Comment by Blinue** [comment]
汇总一下遇到的 XAML Islands 的 bug。（因为 WinUI3 太不给力，我不得不使用这个非常不成熟的技术，权当为后人踩坑）

1. 任务栏上的空窗口，只在 Win10（很可能只在 21H2）上存在
https://github.com/microsoft/microsoft-ui-xaml/issues/6490
![动画](https://user-images.githubusercontent.com/34770031/167993572-d43d13d3-51e5-4158-8c82-03c9ea54b4d3.gif)
这个 bug Windows Terminal 也存在（看来不是我能解决的）
2. Win11 下应用 Mica 时窗口标题渲染不正确（Win11 22H2 中已修复）
![image](https://user-images.githubusercontent.com/34770031/158578704-a504740b-09fb-4067-9380-feafe013da41.png)
3. VS 的 XAML 设计功能无法解析 WinUI，开发体验极差（据说 WinUI 3 也不支持 XAML 设计器）

**Comment by neavo** [comment]
手动点赞支持大佬 👍 另外提一个功能建议：

目前 “DwmSharedSurface” 是不能正常捕获一些画面的（比如 RPGMaker MV 引擎会黑屏），而 "Graphics Capture" 又无法还原被系统缩放过的应用的真实分辨率

导致在实际使用过程中需要频繁的手动切换捕获模式，不知道系统有没有API可以识别目标应用的缩放状态，如果可以根据缩放状态自动选择对应的捕获模式就方便很多了

**Reply by Blinue** [reply]
正在考虑这个功能。DwmSharedSurface 没法捕获到 DirectComposition，但没法判断窗口是否使用了这个技术

**Reply by Blinue** [reply]
我已经决定增加一个“自动”捕获模式，以尽可能的降低使用门槛。

非 DPI 感知或 win10 v2004 以下->GDI
否则->Graphics Capture

理由是：
1. 虽然无法判断窗口是否使用了 DirectComposition，但这类窗口基本都是 DPI 感知的，因此对于非 DPI 感知的窗口可以放心使用 GDI。 DwmSharedSurface 和 GDI 的捕获能力相同，DwmSharedSurface 稍快，但 GDI 兼容性更好。
2. win10 v2004 以下的 Graphics Capture 可能有双光标问题

**Comment by Blinue** [comment]
![动画](https://user-images.githubusercontent.com/34770031/168278176-0d40c7e2-c744-4c07-a2cb-91b21d493bef.gif)

**Reply by Blinue** [reply]
mica 有一些 bug，或许不用更好

**Reply by Blinue** [reply]
Mica 的主要问题已经解决！

**Comment by hooke007** [comment]
shader或者json规则是不是可以考虑加入支持条件限制，看到有人在把fsr用于dscaling，ssim用于upscaling。。。

**Reply by Blinue** [reply]
限制起来代价比较大。暂时不考虑

**Reply by XiaofengdiZhu** [reply]
> shader或者json规则是不是可以考虑加入支持条件限制，看到有人在把fsr用于dscaling，ssim用于upscaling。。。

@Blinue 加上类似madvr配置文件那样的条件判断功能吧，例如仅当原窗口小于等于1080p放大到4k及以上时，该步算法才生效

**Reply by hooke007** [reply]
你说的不就是我说的吗（

**Reply by Blinue** [reply]
这个需求和 #223 类似。这个功能和保存特定游戏配置有很大重合。

**Reply by XiaofengdiZhu** [reply]
> 你说的不就是我说的吗（

给每一步设置条件和给一个算法全局设置限制还是有区别的

**Reply by hooke007** [reply]
那不还是我说的json 或 shader？？？

**Reply by XiaofengdiZhu** [reply]
> 那不还是我说的json 或 shader？？？

算了，你认为没区别就没区别吧

**Reply by hooke007** [reply]
madvr的profile我也在用，这种写写profile的方式完全属于本项目的json的实现范畴，以及#223还扩充了这个需求。
shader限制则类似mpv的glsl WHEN条件语句。

那么你提出的又是额外的哪一种方式？

**Reply by Mishini** [reply]
fsr是基于Lanczos优化的，为什么不能用于dscaling？ssim用于upscaling这个就无语了，电脑性能太强没地方用

**Reply by hooke007** [reply]
> [EASU] Edge Adaptive Spatial Upsampling ....... 1x to 4x area range spatial scaling, clamped adaptive elliptical filter.

如果有进一步质疑可以询问AMD

**Reply by Blinue** [reply]
> fsr是基于Lanczos优化的，为什么不能用于dscaling？

FSR 被设计用于放大，也是针对放大优化的。和它类似的 NIS 用于缩小时会出错。

> ssim用于upscaling这个就无语了，电脑性能太强没地方用

SSIM 用于放大画面也会出错，Magpie 没有机制来限制。

**Reply by Mishini** [reply]
> fsr是基于Lanczos优化的，为什么不能用于dscaling？ssim用于upscaling这个就无语了，电脑性能太强没地方用

我第一次用NIS缩小时出错了还没发现是哪的问题，但是缩放时FSR_EASU比起Lanczos、Bicubic并无明显错误

**Reply by frdoma** [reply]
> > fsr是基于Lanczos优化的，为什么不能用于dscaling？ssim用于upscaling这个就无语了，电脑性能太强没地方用
> 
> 我第一次用NIS缩小时出错了还没发现是哪的问题，但是缩放时FSR_EASU比起Lanczos、Bicubic并无明显错误

细节是要变差一些，虽然不会提示报错，确实不合适用于下采样。

**Comment by Blinue** [comment]
我已经将 MSVC 运行时的链接改为静态链接，虽然会稍微增大程序体积，但不需要用户安装 MSVC 运行时了。

~~v0.10 将不需要用户安装任何运行时，做到开箱即用。~~
v0.10 不能使用这个技术。UWP 应用使用的是特殊版本的 MSVC 运行时，因此微软提供了 [Microsoft.VCRTForwarders.140](https://github.com/microsoft/vcrt-forwarders) 将调用路由到系统的 MSVC 运行时。因为这一次跳转，MSVC 运行时便不支持静态链接了。

**Comment by hooke007** [comment]
LS刚添加了一个窗口模式，虽然有点bug，可以自由拖动放大后的窗口比较有意思

**Reply by Blinue** [reply]
对鼠标的支持怎么样

**Reply by hooke007** [reply]
没有什么太大问题，除了拖动窗口的时候有点延迟/黏手（也就是我说的bug）

**Reply by hooke007** [reply]
![Snipaste_2022-05-19_00-19-25](https://user-images.githubusercontent.com/41094733/169170761-73ffcbdf-68b9-491b-9229-837cce5bf2a8.jpg)

...好吧果然是个实验性功能，刚说完就发现其它bug了（
![Snipaste_2022-05-19_00-18-44](https://user-images.githubusercontent.com/41094733/169170779-59e253ef-9020-4f55-b98f-15379a1394b7.jpg)

**Reply by hooke007** [reply]
这应该和窗口的类型有关系，测试了一下。有的窗口可以触碰到标题栏随意拖动位置，有的被限制在标题栏下方

**Reply by Blinue** [reply]
鼠标没法移出窗口吗

**Reply by hooke007** [reply]
不能。但是如果切出其它界面，它又可以跨窗口交互操作，同时原窗口不会失去缩放。（用的近邻缩放，便于分辨）
![Snipaste_2022-05-19_00-41-11](https://user-images.githubusercontent.com/41094733/169172708-9b4a93d0-257d-40ed-923d-efbb73f4095e.jpg)

**Reply by Blinue** [reply]
很有意思的实现方式。

想完美实现窗口化非常困难。我个人不需要这个功能，但有人认为这个功能很重要，着实鸡肋。我计划在 v0.11 实现窗口化。

（窗口化对 3D 游戏没有意义，LS 基本上不需要它）

**Comment by Blinue** [comment]
计划引入一个“便携模式”，打开时配置文件保存到程序所在目录，否则保存到AppData目录，这样以后更新也可以保留配置了。
默认关闭，否则有逻辑问题

**Comment by Blinue** [comment]
计划的每个页面的功能：

主页：使用方法，激活热键和游戏内覆盖热键，x秒后缩放，自动恢复全屏选项，配置缩放模式预设
缩放配置（全局、单个程序）：缩放模式、捕获模式、各种运行时选项
关于：版本号、更新、许可协议等
选项：主题、便携模式、语言、开机启动、管理员、日志级别、调试

**Comment by MindSerene** [comment]
感谢作者的付出，希望能越做越好。
我根据实列添加了NIS，但是资源占用，功耗增加巨大
我的配置是AMD 5800X   显卡是RX6800， WIN10
原生1080永劫无间340FPS，放大到4k120hz画面，，用LS的NIS放大后是325 FPS
用Magpie放大后是140FPS，占用资源超越了原生4K，性能反而下降了。
换到WIN11后，Graphics Capture捕获模式无法正常使用，要更换Desktop Duplication模式。
而LS没有模式选项，无法在WIN11下使用。
是什么原因导致了Magpie的效率大打折扣呢，我挺不理解的。
配置文档是：
    "name": "NIS",
    "effects": [
      {
        "effect": "NIS",
        "scale": [ 2, 2 ],
        "sharpness": 0.3

**Reply by Blinue** [reply]
能贴下日志吗

**Reply by hooke007** [reply]
> [logs.zip](https://github.com/Blinue/Magpie/files/8790947/logs.zip)
> 是这个么，看了下21号之后内容已经是WIN11了。

帮你把帖子日志移动过来了，没必要新开楼层（

**Reply by Blinue** [reply]
你是否启用了垂直同步？在全屏下按下 Ctrl+Alt+D 可显示游戏内覆盖，可以便捷查看 Magpie 的性能。

> 永劫无间不是游戏内集成nis么

游戏内置的缩放功能永远是首选，其次是驱动提供的缩放功能，最后才是 Magpie/LS。我从来不推荐使用 Magpie 来提高游戏性能。

**Reply by MindSerene** [reply]
> 永劫无间不是游戏内集成nis么

内置的是DLSS，需要30系列显卡，我是RX6800的A卡

**Reply by MindSerene** [reply]
> 你是否启用了垂直同步？在全屏下按下 Ctrl+Alt+D 可显示游戏内覆盖，可以便捷查看 Magpie 的性能。
> 
> > 永劫无间不是游戏内集成nis么
> 
> 游戏内置的缩放功能永远是首选，其次是驱动提供的缩放功能，最后才是 Magpie/LS。我从来不推荐使用 Magpie 来提高游戏性能。

并不是拿来提升游戏性能，而是一种很优秀的应用。
我是用NIS来把1080P的画面铺满4K并且达到接近原生4K的画质，关键它只是用额外付出10~20%性能来获得的。
例如：同样锁定120FPS的画面，原生1080P使用游戏自带的无边框全屏化，功耗70W，略糊，原生4k的话功耗185W
在LS下用NIS放大后功耗80W，画质很接近原生4k
，而用Magpie一样的1K放大到4K拉满了显卡功耗190W
这就是奇怪的地方了。
<img src="https://user-images.githubusercontent.com/86352601/170860756-729230b6-7eb6-4e78-96b3-1c3a16eda0f0.png" width=500>

**Reply by MindSerene** [reply]
> 你是否启用了垂直同步？在全屏下按下 Ctrl+Alt+D 可显示游戏内覆盖，可以便捷查看 Magpie 的性能。
> 
> > 永劫无间不是游戏内集成nis么
> 
> 游戏内置的缩放功能永远是首选，其次是驱动提供的缩放功能，最后才是 Magpie/LS。我从来不推荐使用 Magpie 来提高游戏性能。

我没有开启Magpie的垂直同步，游戏的垂直同步也关闭了，帧率限制120,游戏内也是120FPS
但是按Ctrl+Alt+D看到的AVG fps是游戏的10倍（1100~1200 fps），这正常吗

**Reply by hooke007** [reply]
![Snipaste_2022-05-29_10-19-00](https://user-images.githubusercontent.com/41094733/170861010-5c0c49fd-2a8f-4b83-8e39-c350091c6f57.png)
这游戏去年的更新日志里就集成nis了

**Reply by Blinue** [reply]
> 我没有开启Magpie的垂直同步，游戏的垂直同步也关闭了，帧率限制120,游戏内也是120FPS

你在 Magpie 中关闭了垂直同步，因此 Magpie 的帧率是无限制的，即会使用所有的显卡资源，功耗当然会很高。应始终开启垂直同步。

**Reply by hooke007** [reply]
> <img src="https://user-images.githubusercontent.com/96861853/170861227-dd70e08e-6713-462a-9913-1ef2a0aa1bcb.jpg" width=500>
> 这个图形增强不就是nis么  你玩这个游戏竟然不知道? 当时集成nis后甚至还有一定热度
@decadeicey 原层合并到这里

**Reply by MindSerene** [reply]
> 

！原来如此，这就难怪了。开启后确实没有占用那么高的资源了。这个有点儿隐蔽啊，因为游戏设置了垂直同步，根本看不出来
这样功耗确实正常了。也是80W了，而且画质是一样的，
我尝试了游戏自带的 设置51%渲染激活NIS，全屏4K，功耗倒也是80W，但是画面并没有第三方NIS的清晰，不过比1k无边框全屏好

用NIS放大1080P的视频到4k，也能清晰不少，达到伪4K画质，对动漫尤其有效，我习惯了这办法后爱不释手。
再次感谢作者以及其他协助者！

**Reply by MindSerene** [reply]
> > 我没有开启Magpie的垂直同步，游戏的垂直同步也关闭了，帧率限制120,游戏内也是120FPS
> 
> 你在 Magpie 中关闭了垂直同步，因此 Magpie 的帧率是无限制的，即会使用所有的显卡资源，功耗当然会很高。应始终开启垂直同步。

目前NIS使用正常了，比游戏自带的画质要好，估计游戏本身没有设置锐化？（我设置的3~5）

建议Magpie内置一个功能，放大之前设置源窗口分辨率大小，不然一些自由缩放的窗口固定放大2：2后不太对劲
比如用bilibili的客户端版本用小窗播放视频的时候，要先把窗口手动拖到1920*1080，再放大才是完全正常的
这功能在LS里是有的……在Magpie找不到，找到的是裁剪，似乎不是这样用的

**Reply by Blinue** [reply]
将 scale 参数设为 [-1,-1] 即可适应屏幕

**Reply by hooke007** [reply]
> 先把窗口手动拖到1920*1080，再放大

这个 FR 有人提过了 https://github.com/Blinue/Magpie/issues/173

**Reply by MindSerene** [reply]
> 将 scale 参数设为 [-1,-1] 即可适应屏幕

设置了，效果并不好，比如视频窗口比较小或者不是16：9，放大后的效果就不合适，
手动拖成1920*1080采取Desktop Duplication 模式放大（WIN11似乎只能用这个），
会在屏幕最边缘露出窗口外的底色，比如桌面壁纸的绿色之类的（自定义裁剪掉四周2像素似乎也可以）
这方面还是缩放前调整到指定大小最好。
LS没这方面细节问题，不过LS在WIN11下放大bilibili播放器的话要设置bilibili播放器禁用GPU加速，否则黑屏。
它可能就是采取的另一个抓取模式吧。放大后没边框线。

**Reply by MindSerene** [reply]
> > 先把窗口手动拖到1920*1080，再放大
> 
> 这个 FR 有人提过了 #173

你的记忆力可真惊人。
这个功能在LS中用的很顺手，一些自由拉伸并且不可设置分辨率的窗口非常需要它。

**Reply by Blinue** [reply]
关于调整源窗口大小的功能请前往 #173 讨论。

Graphics Capture 是兼容性最好的捕获模式，在 Win11 上应该不会有失败的情况。如果你遇到了，请发下日志。边线是小窗自己绘制的，只能通过裁剪解决。LS 的捕获模式 Magpie 也有，选择 GDI 或 DwmSharedSurface。

**Reply by MindSerene** [reply]
> 关于调整源窗口大小的功能请前往 #173 讨论。
> 
> Graphics Capture 是兼容性最好的捕获模式，在 Win11 上应该不会有失败的情况。如果你遇到了，请发下日志。边线是小窗自己绘制的，只能通过裁剪解决。LS 的捕获模式 Magpie 也有，选择 GDI 或 DwmSharedSurface。


调用  Graphics Capture  后Magpie会留下log后消失，没任何画面变化。log里我似乎看不出什么，最前面就是崩溃的时候留下的。
[logs.zip](https://github.com/Blinue/Magpie/files/8793122/logs.zip)
Desktop Duplication 比较正常，
另外两个模式黑屏。（bilibili播放器启用GPU加速的情况下,LS也是黑的）
关闭播放器的GPU加速后，Magpie 的GDI模式似乎就是LS的抓取模式，Magpie 抓取也没有边线了。
DwmSharedSurface模式花屏，类似反色，全是红绿紫之类的颜色……

**Reply by Blinue** [reply]
> 调用 Graphics Capture 后Magpie会留下log后消失，没任何画面变化。log里我似乎看不出什么，最前面就是崩溃的时候留下的。

闪退了的话就没办法了，没有日志供分析。

> DwmSharedSurface模式花屏，类似反色，全是红绿紫之类的颜色……

和 https://github.com/Blinue/Magpie/issues/94#issuecomment-933184118 情况一样，是驱动配置问题。（GDI 兼容性比 DwmSharedSurface 好，这也是意料之中）

**Reply by MindSerene** [reply]
> > 调用 Graphics Capture 后Magpie会留下log后消失，没任何画面变化。log里我似乎看不出什么，最前面就是崩溃的时候留下的。
> 
> 闪退了的话就没办法了，没有日志供分析。
> 
> > DwmSharedSurface模式花屏，类似反色，全是红绿紫之类的颜色……
> 
> 和 [#94 (comment)](https://github.com/Blinue/Magpie/issues/94#issuecomment-933184118) 情况一样，是驱动配置问题。（GDI 兼容性比 DwmSharedSurface 好，这也是意料之中）

同样的硬件配置在WIN10下似乎不会闪退，需要多收集些其他人的情况才能慢慢找到原因了吧。
前面提到的永劫无间自带的NIS估计是魔改过的吧？
它能保持窗口化调用NIS，而不像Magpie这样必须全屏调用，
这是不是说明了不全屏而指定放大后依然是窗口分辨率的可能性？

**Reply by Blinue** [reply]
NIS 是缩放算法，和是否全屏无关。窗口化问题见 #135

**Reply by MindSerene** [reply]
> NIS 是缩放算法，和是否全屏无关。窗口化问题见 #135

今天开始我看哔哩哔哩视频采用了Magpie了，有一个建议，功能开放可选：是否限制鼠标在窗口范围内。
这功能对玩游戏很有用，是必需品，但是对视频可能不是，因为放大后鼠标可能会影响观看，会触发播放器UI。

**Reply by Blinue** [reply]
看视频有其他更好的方式...

这个需求太小众，不会实现

**Reply by MindSerene** [reply]
> 看视频有其他更好的方式...
> 
> 这个需求太小众，不会实现

AMD在2022年6月20日更新了WIN11的显卡驱动22.5.2的Optional (WHQL)版本
之前Magpei在我这不能用Graphics Capture模式的问题修复了，现在我可以使用这个模式了。
另外说一个和Magpie可能没有直接关系的问题，流放之路设置为窗口无边框全屏模式的时候，载入游戏过程，用Magpie放大另一个窗口，流放之路可能会以另一个窗口所在桌面的坐标为全屏的起点，就会很奇怪的空着桌面的左上角（我这大概是四分之一屏幕），游戏又是全屏状态。类似你手动拖动一个全屏窗口到了那个位置，而这操作原来是不可行的。

**Comment by Blinue** [comment]
Windows 有一个恼人的 bug，即使在窗口类中设置了背景颜色，在窗口启动时背景却是白色，造成瞬间的闪烁。在暗色背景下非常明显

![动画](https://user-images.githubusercontent.com/34770031/171983568-8c892b87-56af-4571-9b42-398ae86b206e.gif)

一个 workaround：在调用 ShowWindow 之前先执行
```
SetWindowPos(hWnd, NULL, 0, 0, 0, 0, SWP_NOMOVE | SWP_NOSIZE | SWP_NOREDRAW);
```
这似乎可以强制刷新背景。现在效果完美了：

![动画](https://user-images.githubusercontent.com/34770031/171983330-3295664d-cc88-4db8-9e9f-fe3f21ff4c89.gif)

Stack Overflow 中有这个 bug 的讨论：https://stackoverflow.com/questions/69715610/how-to-initialize-the-background-color-of-win32-app-to-something-other-than-whit

**Comment by Blinue** [comment]
一直以来使用 [globalmousekeyhook](https://github.com/gmamaladze/globalmousekeyhook) 实现热键功能，但 v0.10 将自行实现热键（使用 [RegisterHotKey](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-registerhotkey)）。更改热键的方法也会优化，和 PowerToys 保持一致。

热键将由两部分组成：Modifiers + Virtual Key

Modifiers ：Windows、Ctrl、Alt、Shift 的组合
Virtual Key：支持的键位有：字母、数字（顶部和小键盘）、回车、空格、PageUp、PageDown、End、Home、方向键、Insert、Delete、加减乘除、`、F1~F12、逗号、句号、分号、分隔符、[、]、\、-、=、Backspace

**支持不指定 Virtual Key**

使用 RegisterHotKey 的额外好处是，它可以检测冲突的热键，因此至少可以确保不会和系统热键重复。

**Comment by Blinue** [comment]
**2022/7/10 更新**

决定始终匹配完整路径+窗口类名，这是为了便于检索图标。

UWP 窗口（Packaged Apps）则匹配 PackageFamliyName + Package-Relative App Identifier (PRAID) + 窗口类名，这方面很复杂。

-----

识别窗口是一个技术活。我可以保存针对某个窗口的配置，但我怎样才能在以后识别出同一个窗口？

我查阅了 Borderless Gaming 的源码，发现它的逻辑异常简单，只是匹配标题，如果标题为空，就匹配可执行文件名。我还尝试了 Xbox Game Bar，发现基本上也是匹配标题。匹配标题非常不可靠，永远没法适配记事本。

没找到可以可靠识别窗口的方法，因此这方面将由用户决定。将提供四个匹配项：

1. 可执行文件路径
2. 可执行文件名
3. 窗口类名
4. 窗口标题

通过不同方法添加的配置将有不同的默认值：

1. 可以通过指定可执行文件路径添加配置，此时默认值为 1、2
2. 可以通过指定现有窗口添加配置，此时默认值为 2、3
3. 可以通过游戏内覆盖保存配置，此时默认值为 2、3

一些使用方法：

1. 同时指定可执行文件路径和可执行文件名可锁定特定文件，如果想要移动位置后依然可用，不要指定可执行文件路径
2. 大多数情况下 23 就够用了，如果窗口标题不会改变，推荐也指定 4

**Reply by neavo** [reply]
> 识别窗口是一个技术活。我可以保存针对某个窗口的配置，但我怎样才能在以后识别出同一个窗口？
> 
> 我查阅了 Borderless Gaming 的源码，发现它的逻辑异常简单，只是匹配标题，如果标题为空，就匹配可执行文件名。我还尝试了 Xbox Game Bar，发现基本上也是匹配标题。匹配标题非常不可靠，永远没法适配记事本。
> 
> 没找到可以可靠识别窗口的方法，因此这方面将由用户决定。将提供四个匹配项：
> 
> 1. 可执行文件路径
> 2. 可执行文件名
> 3. 窗口类名
> 4. 窗口标题
> 
> 通过不同方法添加的配置将有不同的默认值：
> 
> 1. 可以通过指定可执行文件路径添加配置，此时默认值为 1、2
> 2. 可以通过指定现有窗口添加配置，此时默认值为 2、3
> 3. 可以通过游戏内覆盖保存配置，此时默认值为 2、3
> 
> 一些使用方法：
> 
> 1. 同时指定可执行文件路径和可执行文件名可锁定特定文件，如果想要移动位置后依然可用，不要指定可执行文件路径
> 2. 大多数情况下 23 就够用了，如果窗口标题不会改变，推荐也指定 4

😂 2+3 应该能区分不同代数的RPGMaker的game.exe吧

**Reply by Blinue** [reply]
游戏很少有改标题的，但有的游戏会把章节放在标题里。

**Reply by Blinue** [reply]
所有 UWP 都使用同一个可执行文件，难顶

**Comment by Blinue** [comment]
https://github.com/ActualMandM/cemu_graphic_packs/blob/master/Filters/Bicubic/output.glsl

一个很快的 Bicubic 实现

**Reply by Blinue** [reply]
Bicubic 处理时间减少 20%，比 CatmullRom 还快了

CatmullRom 不再有性能优势，将在 v0.10 删除

**Comment by Blinue** [comment]
正在考虑一个新功能，在保持鼠标流畅的同时降低功耗。比如虽然以 60 FPS 运行，但画面刷新率只有 30 FPS，间隔的一帧只更新鼠标。

**Reply by XiaofengdiZhu** [reply]
是指**Magpie缩放后画面**以60fps运行，但**原本游戏**画面刷新率只有30fps吗？我这有个2fps的游戏，鼠标卡飞……

**Reply by Blinue** [reply]
缩放画面 60 FPS，以 30 FPS 捕获和放大

**Reply by XiaofengdiZhu** [reply]
> 缩放画面 60 FPS，以 30 FPS 捕获和放大

挺好，这样2fps的游戏放大后鼠标就不会卡飞了

**Reply by Blinue** [reply]
光标卡顿取决于 Magpie 本身的帧率。如果你没有使用重量级的缩放算法，那么可能是因为游戏本身耗尽了显卡资源，导致 Magpie 帧率也很低。

**Reply by XiaofengdiZhu** [reply]
> 光标卡顿取决于 Magpie 本身的帧率。如果你没有使用重量级的缩放算法，那么可能是因为游戏本身耗尽了显卡资源，导致 Magpie 帧率也很低。

amd核显用graphics capture模式时，如果游戏输出2帧，那么Magpie输出的也是2帧，非常非常节能

**Comment by Blinue** [comment]
经测试，mega/xaml 分支与 RTSS 兼容。https://github.com/Blinue/Magpie/issues/202#issuecomment-983685195

**Comment by Blinue** [comment]
WinUI 页面间导航存在内存泄露问题 https://github.com/microsoft/microsoft-ui-xaml/issues/934

WinUI 存在大量的 bug，微软基本上拒绝修复，我用了很多 hack 给它擦屁股

**Comment by Blinue** [comment]
Nearest 和 Bicubic 将内置，防止用户修改或删除。

Nearest：将有一个无法删除的缩放模式 Origin 作为最后回落，该缩放模式使用 Nearest 将捕获内容原样输出。
Bicubic：将有一个“默认降采样算法”选项，当输出尺寸大于屏幕尺寸将自动应用该效果。此选项的默认值为 Bicubic。

**Comment by Blinue** [comment]
为了支持用户界面，MagpieFX 需要更新：

1. 支持以文件夹形式分组而不是必须放在 effects 根目录
2. 参数的 DEFAULT、MIN、MAX、STEP 都必须存在，可选 LABEL，将基于它们生成用户界面
3. GENERIC_DOWNSCALER 表示此效果是通用的降采样算法，这用于实现“默认降采样效果”功能。此标志和 OUTPUT_WIDTH/OUTPUT_HEIGHT 冲突
4. BUILT_IN 表示内置效果，自动更新时将被删除。自定义效果不要使用此标志

**Comment by Blinue** [comment]
发现了一个控制 GPU 功耗的接口 [GPU Performance State Interface](https://docs.nvidia.com/gameworks/content/gameworkslibrary/coresdk/nvapi/group__gpupstate.html)。不知道 AMD 有没有类似的东西

https://1vwjbxf1wko0yhnr.wordpress.com/

**Comment by Blinue** [comment]
目标 SDK 已升级为 22621

**Reply by Blinue** [reply]
22621 对 Mica 有完美支持，可以不用自己实现了

**Comment by Blinue** [comment]
自动更新时自定义效果文件怎么处理是个问题

**Reply by Blinue** [reply]
为效果增加 BUILT_IN 标志，含有此标志的效果将在自动更新时被删除

**Comment by Mr-Z-2697** [comment]
0.10.0-preview1 重命名以及新建缩放模式配置的时候，输入框在大部分情况下无响应（可以粘贴，可以点×清除，谷歌拼音输入法2可以输入）
系统版本：Windows 10 22H2 19045.2251

**Reply by Blinue** [reply]
我在 Win10 22H2 上没能复现，肯定有某种触发条件

**Reply by Mr-Z-2697** [reply]
我又试了一下，重新开启之后是没有问题的，但是关闭到托盘之后再打开主界面就会出现问题

**Reply by Blinue** [reply]
已经复现，似乎只出现在win10上

**Reply by Blinue** [reply]
虽然不知道为什么，但最新的预览版似乎没有这个 bug。你试一试
->[点此下载](https://github.com/Blinue/Magpie/suites/9462665196/artifacts/448017455)

**Reply by Blinue** [reply]
更多信息：在 https://github.com/Blinue/Magpie/commit/21fe1b01a9572f152433b0c8386b97987be14804 里这个 bug 被修复。在这个提交里，用于接收托盘图标消息的隐藏窗口由 message-only 变为常规窗口。原理无从知晓，XAML Islands 是一个神秘的系统。

**Reply by Mr-Z-2697** [reply]
确实没问题了。。。

**Comment by Mr-Z-2697** [comment]
另外，在配置的时候有没有办法增加直接替换的选项？否则要替换某个效果的话只能删除、添加、拖动排序……

**Reply by Blinue** [reply]
决定放在右键菜单里

**Comment by Blinue** [comment]
突然发现 magpie 已经可以通过 Chocolatey 安装了。感谢维护者。
https://community.chocolatey.org/packages/magpie

**Comment by Blinue** [comment]
检查更新流程
```mermaid
flowchart LR
    Pending-- 手动或自动 -->Checking
    Checking<-->NoUpdate
    Checking<-->Available
    Checking<-->ErrorWhileChecking
    Available-->Downloading
    Downloading<-->ErrorWhileDownloading
    Downloading-->Installing
    NoUpdate-- 进入关于页面或取消 -->Pending
    ErrorWhileChecking-- 进入关于页面或取消 -->Pending
    Available-- 关闭主窗口或取消 -->Pending
    Downloading-- 取消 -->Available
    ErrorWhileDownloading-- 取消 -->Pending
```

**Comment by hooke007** [comment]
我看了一下好像没人反馈gsync的问题。
现在gsync会正常应用在放大后的窗口（GC）上，但是这引入了鼠标不跟手（先停止移动光标，然后突然移动，明显的瞬间卡顿）的问题。
启用magpie的“垂直同步”选项并没有效果，可以观察到静止时fps降到了1，当前可用的解决方法是禁用direct-flip

**Reply by XiaofengdiZhu** [reply]
fps为1说明游戏节能，Magpie此时也节能，不开gsync也会鼠标卡，我建议Magpie加一个判断：如果鼠标在窗口上移动，则尽可能输出最大帧率（或指定帧率）

**Reply by hooke007** [reply]
如果不开gsync也卡——说明你的问题不是我说的问题，我描述的是开启gsync才出现的特有问题，这里不开gsync一切正常

**Reply by Blinue** [reply]
GC 的帧率由系统控制，如果窗口内容不变鼠标也没移动就会降低帧率。或许可以用一些 hack 阻止 GC 降低帧率，禁用 DirectFlip 是可行的，但牺牲了延迟和画面质量。

**Comment by Blinue** [comment]
自动更新 #287 已经实现🎉
因为使用 github 源，国内用户大概很难成功。

**Reply by BLACKCHARGE** [reply]
如果支持手动更换镜像源的功能估计会好点

**Reply by Blinue** [reply]
我不想做的太复杂，维护麻烦。更新功能会自动使用系统代理。

**Comment by Blinue** [comment]
Magpie 已引入 [Hybrid CRT](https://github.com/microsoft/WindowsAppSDK/blob/main/docs/Coding-Guidelines/HybridCRT.md)。

桌面应用程序通常通过两种方式链接 VCRT，其中动态链接需要用户本地安装 MSVC 运行时，但产生的可执行文件较小，静态链接将运行时作为静态库链接到可执行文件，因此无需用户安装环境，但产生的可执行文件体积较大。

Hybrid CRT 可视为两者的折中，它的原理在于：Win10/11 中已经部署了 UCRT (Universal CRT)，它是 VCRT 的子集。使用 Hybrid CRT 
 时，对于 UCRT 中存在的 API （如 C 运行时库），将动态链接，而不存在的 API（如 STL）则静态链接。因此 Hybrid CRT 无需用户安装 MSVC 运行时，又不会使可执行文件的体积膨胀太多。

Hybrid CRT 目前只在 WindowsAppSDK 中使用，但实际上可以移植到任何桌面应用程序。

下面是 Magpie 使用不同链接技术时的体积：
| | Dynamic | Hybrid | Static |
| --- | --- | --- | --- |
| Magpie.exe | 609K | 646K | 838K |
| Magpie.App.dll | 2873K | 2920K | 3113K |
| Magpie.Core.dll | 1349K | 1607K | 1880K |
| Updater.exe | 86K | 114K | 190K |

对于 Magpie 来说，迁移到 Hybrid CRT 的主要优势在于易于部署。还有两个小问题，我认为是可以接受的：
1. mimalloc 只支持动态链接 VCRT，因此不能使用了。
2. Conan 引入的依赖无法使用 Hybrid CRT（我还没找到办法），不过其中最常使用的几个都是 header-only 的。

**Comment by Blinue** [comment]
自动缩放 #193 已经实现🎉

**Comment by Blinue** [comment]
我已经开始实现本地化 #476 。可行的话会使用 https://hosted.weblate.org ，它对自由软件似乎有免费额度。

**Reply by Blinue** [reply]
使用 XAML Islands 时，在运行时切换语言是可能的：
``` c++
ResourceContext::ResetGlobalQualifierValues({ L"Language" });
ResourceContext::SetGlobalQualifierValue(L"Language", tag);
```
然后重新加载整个用户界面即可。

但我决定不支持这个功能，也不响应操作系统的语言更改。一方面切换语言很少使用，另外支持在运行时切换语言会增加用户界面的复杂度，拖累速度，以及增加维护的成本。

为了切换语言，用户需要重启 Magpie。

**Reply by Blinue** [reply]
我几天之前申请了 Weblate 的自由软件项目托管，可惜到现在还没回应。如果成功，以后就会用 Weblate 贡献翻译

https://hosted.weblate.org/projects/magpie/

**Reply by Blinue** [reply]
托管申请最后都没有回复😅
![未命名1678238447](https://user-images.githubusercontent.com/34770031/223891260-16b44350-7eb6-44a7-8b64-bbbf747d1e42.png)

现在项目被删除了，我要想其他办法了

**Reply by Blinue** [reply]
打算用 https://crowdin.com

**Comment by Blinue** [comment]
搞了一个捐助渠道。因为我没有 PayPal，国外用户就不支持了。

**Comment by Blinue** [comment]
一个好消息，我的开发环境鸟枪换炮了，所以以后会有更多针对高配置的优化，包括高分辨率、高刷新率、HDR、VRR 等。

**Reply by XiaofengdiZhu** [reply]
3090换h100，xmsl（逃

**Reply by Blinue** [reply]
等我有钱了一定搞一个 h100，这样你们遇到性能问题我都能说：无法复现🤣
